### PR TITLE
feat: added notation cert cleanup-test command

### LINKED ISSUE #647: Support clean up the source key and certificate generated by Notation
### What is the areas you would like to add the new feature to?

Notation CLI

### Is your feature request related to a problem?

`notation key delete` can only remove the key from the signing key list and `notation cert delete` can only remove the self-signed certificate from the trust store. This is by design since Notation doesn't support signing with [local keys](https://github.com/notaryproject/notation/issues/539) and managing local keys. 

Per discussion in https://github.com/notaryproject/notation/pull/606#discussion_r1154664392 and another issue #604 , users want to delete the source key and certificate generated by `notation cert generate-test` in a convenient way.  

### What solution do you propose?

Providing a flag `--cleanup` to `notation cert generate-test` to allow users to delete the specified source key and certificate generated by `notation cert generate-test`. This flag is only used to delete the test key and self-signed certificate. The keys and certificates that are not generated by Notation will not be able to be deleted with this flag. 

For example, delete a source key and cert generated by `notation cert generate-test "wabbit-networks.io"`:

```
$ notation cert generate-test --cleanup wabbit-networks.io
Deleted <key_name> and <cert_name> 
```

### What alternatives have you considered?

N/A

### Any additional context?

_No response_

### PR DESCRIPTION
**Pull request by Two-Hearts**
In this PR:
1. Adds `notation cert cleanup-test` command, this is to cleanup key-cert pairs created by `notation cert generate-test`. On success, both key and cert files are deleted; the self-signed cert is removed from trust store; the private key is removed from notation's signing key list.
2. This new command is behind `experimental` at the moment, because we may support signing with local keys in the future. Open to suggestions about whether we should move it out of 'experimental'.

Resolves #647.

### CODE REVIEW

File: cmd/notation/cert/cleanupTest.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+			return errors.New("cleanup-test can only apply to non-external keys created by the generate-test command")
+		}
+		if keySuite.X509KeyPair == nil {
+			return errors.New("cleanup-test requires key pair files created by the generate-test command")
+		}
+		// delete the key file from keyPath
+		keyPath := keySuite.X509KeyPair.KeyPath
+		err = osutil.DeleteFile(keyPath)
+		if err != nil {
+			finalError = append(finalError, fmt.Errorf("cannot delete the key file from %q during cleanup-test, %v", keyPath, err))
```
Comment (JeyJeyGao) [inline cmd/notation/cert/cleanupTest.go:58]: Why not print the error directly?

File: internal/osutil/file_test.go [hunk 2]
```diff
@@ -2,19 +2,18 @@ package osutil
...
-	"io/ioutil"
 	"os"
 	"path/filepath"
 	"runtime"
 	"testing"
 )
 
 func validFileContent(t *testing.T, filename string, content []byte) {
-	b, err := ioutil.ReadFile(filename)
+	b, err := os.ReadFile(filename)
```
Comment (JeyJeyGao) [inline internal/osutil/file_test.go:12]: Is this the unit test for ` ioutil.ReadFile`?

File: cmd/notation/cert/cleanupTest.go (hunk 1 above)
Comment (JeyJeyGao) [reply cmd/notation/cert/cleanupTest.go:58]: We may don't need to print an error if the file doesn't exist.

Comment (Two-Hearts) [reply cmd/notation/cert/cleanupTest.go:58]: This PR and its related issue are pending discussions. Will address comments once everyone agrees on the design.

File: cmd/notation/cert/generateTest.go [hunk 3]
```diff
@@ -40,20 +40,20 @@ func certGenerateTestCommand(opts *certGenerateTestOpts) *cobra.Command {
 	}
 	command := &cobra.Command{
 		Use:   "generate-test [flags] <common_name>",
-		Short: "Generate a test RSA key and a corresponding self-signed certificate.",
+		Short: "[Experimental] Generate a test RSA key and a corresponding self-signed certificate.",
```
[the remark this answers is not in the document]
Comment (sajayantony) [reply cmd/notation/cert/generateTest.go:43]: +1 mainline scenarios for developing or testing should not be experimental.

File: cmd/notation/cert/cleanupTest.go (hunk 1 above)
Comment (Two-Hearts) [reply cmd/notation/cert/cleanupTest.go:63]: As we discussed offline, there are, in total, 4 deletions/removals in cleanup-test. We want all deletions/removals to be run even some of them may fail during this process. Eventually, we will print out all the succeeded operations followed by all the failed cases. If we return the error directly, then the user may need to call `notation cert cleanup-test` multiple times to remove all the related files which is not very user friendly.

File: cmd/notation/cert/generateTest.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Two-Hearts) [reply cmd/notation/cert/generateTest.go]: @priteshbandi @sajayantony Got it. I updated the code to reflect your suggestions.

File: cmd/notation/cert/cleanupTest.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+type certCleanupTestOpts struct {
+	keyName string
+}
+
+func certCleanupTestCommand(opts *certCleanupTestOpts) *cobra.Command {
+	if opts == nil {
+		opts = &certCleanupTestOpts{}
+	}
+	command := &cobra.Command{
+		Use:   "cleanup-test <common_name>",
```
Comment (shizhMSFT) [inline cmd/notation/cert/cleanupTest.go:24]: It seems the `notation cert cleanup-test` command is not behind an experimental flag yet.

File: test/e2e/suite/scenario/quickstart.go [hunk 5]
```diff
@@ -22,7 +22,10 @@ var _ = Describe("notation quickstart E2E test", Ordered, func() {
 		if err != nil {
 			panic(err)
 		}
-		vhost.SetOption(AuthOption("", ""))
+		vhost.SetOption(
+			AuthOption("", ""),
+			EnableExperimental(),
```
Comment (shizhMSFT) [inline test/e2e/suite/scenario/quickstart.go:27]: Given `notation cert generate-test` is not an experimental feature as discussed in the above comments by others, do should we still need to `EnableExperimental()`?

File: test/e2e/suite/command/sign.go [hunk 6]
```diff
@@ -66,7 +66,7 @@ var _ = Describe("notation sign", func() {
 	})
 
 	It("with specific key", func() {
-		Host(BaseOptions(), func(notation *utils.ExecOpts, artifact *Artifact, vhost *utils.VirtualHost) {
+		Host(BaseOptionsWithExperimental(), func(notation *utils.ExecOpts, artifact *Artifact, vhost *utils.VirtualHost) {
```
Comment (shizhMSFT) [inline test/e2e/suite/command/sign.go:69]: same here

File: cmd/notation/cert/cmd.go [hunk 7]
```diff
@@ -1,6 +1,8 @@
 package cert
 
-import "github.com/spf13/cobra"
+import (
+	"github.com/spf13/cobra"
+)
```
Comment (shizhMSFT) [inline cmd/notation/cert/cmd.go:5]: nit: should revert the change

File: cmd/notation/cert/cleanupTest.go [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+			return errors.New("cleanup-test requires key pair files created by the generate-test command")
+		}
+		// delete the key file from keyPath
+		keyPath := keySuite.X509KeyPair.KeyPath
+		err = osutil.DeleteFile(keyPath)
+		if err != nil {
+			if os.IsNotExist(err) {
+				fmt.Printf("Success: key file %q does not exist. It is already been cleaned up.\n", keyPath)
+			} else {
+				finalError = append(finalError, fmt.Errorf("cannot delete the key file %q during cleanup-test, %v", keyPath, err))
```
Comment (shizhMSFT) [inline cmd/notation/cert/cleanupTest.go:66]: nit: instead of having a `finalError` slice. Should we print the error directly?
```suggestion
				fmt.Fprintf(os.Stderr, "Failed: cannot delete the key file %q during cleanup-test: %v\n", keyPath, err))
```

Comment (Two-Hearts) [reply cmd/notation/cert/cleanupTest.go:66]: > nit: instead of having a `finalError` slice. Should we print the error directly?

@shizhMSFT I think the readability of the current printouts is higher. All succeeded cases are grouped at the top followed by all the failed cases grouped together as well. In this way, the printouts are very clear and easy to read. What do you think? 
/cc: @yizha1 @FeynmanZhou

File: cmd/notation/cert/cleanupTest.go (hunk 4 above)
Comment (Two-Hearts) [reply cmd/notation/cert/cleanupTest.go:24]: > It seems the `notation cert cleanup-test` command is not behind an experimental flag yet.

Yes, this is something that we need to discuss. Since we decided to not put `notation cert generate-test` behind `experimental`, I think we shouldn't put `notation cert cleanup-test` behind the flag either? This is to make the user experience more consistent before and after.

File: cmd/notation/cert/cleanupTest.go (hunk 8 above)
Comment (shizhMSFT) [reply cmd/notation/cert/cleanupTest.go:66]: I think the output should reflect the order of the execution.

File: cmd/notation/cert/cleanupTest.go (hunk 4 above)
Comment (shizhMSFT) [reply cmd/notation/cert/cleanupTest.go:24]: @Two-Hearts @yizha1 @FeynmanZhou Let's discuss it in the community meeting.

**Comment by Two-Hearts** [comment]
Converted to draft until we are ready to discuss this again.
